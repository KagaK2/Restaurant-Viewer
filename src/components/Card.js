import React from 'react';
import "../styles/Card.scss";


//This component receive a restaurant object as a prop and returns a card with the restaurant's inforamtion.
const Card = ({restaurant}) => {

    const {name, description, image, delivery_price, currency, tags} = restaurant;

    return (
       <div className="Card">

           <div className="image-container">
            <img src={image} alt={name} />
           </div>

           <div className="card-content">
               <div className="card-content-left top"> 
                    <h3>{name}</h3>
                </div>

                <div className="card-content-right top">
                    <h4>{`${delivery_price/100} ${currency}`}</h4>
                </div>

                <div className="card-content-left top">
                    <p>{description}</p>
                </div>

                <div className="card-content-right bottom">
                    <p>{tags.length > 0 && `#${tags.join(", #")}`}</p> 
                </div>
           </div>
       </div>
    )
};

export default Card;