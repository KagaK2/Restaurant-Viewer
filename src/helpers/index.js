//This is a helper file, which contains some helper functions.

// This function takes an array of restaurants and sort it in ascending alphabet order.
export function sortRestaurantsAsc(restaurants){
    //Copy a new array to avoid mutation.
    const copied = restaurants.concat();
    copied.sort((a, b) => { 
        if(a.name.toUpperCase() < b.name.toUpperCase())
            { return -1;} 
        if(a.name.toUpperCase() > b.name.toUpperCase())
            { return 1;} 
        return 0;
    })
    return copied;
}

// This function takes an array of restaurants and sort it in descending alphabet order.
export function sortRestaurantsDes(restaurants){
    //Copy a new array to avoid mutation.
    const copied = restaurants.concat();
    copied.sort((a, b) => { 
        if(a.name.toUpperCase() < b.name.toUpperCase())
            { return 1;} 
        if(a.name.toUpperCase() > b.name.toUpperCase())
            { return -1;} 
        return 0;
    })
    return copied;
}