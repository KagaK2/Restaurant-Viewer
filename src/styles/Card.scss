.Card {
    border-radius: 3%;
    background-color: #f2f2f2;
    max-width: 800px;
    height: 100%;
    overflow: hidden;
    box-shadow: 0 .25rem .25rem rgba(0,0,0,0.2),0 0 1rem rgba(0,0,0,0.2);
    & > .image-container {
        width: 100%;
        overflow: hidden;
    
        & > img {
            width: 100%;
            height: 15vmax;
            object-fit:cover;
        }
    }
    & > .card-content {
        width: 90%;
        height: 30%;
        margin: 0 1rem;
        display: grid;
        grid-template-columns: 65% auto;
        grid-template-rows: 1fr 1fr;
        
        & .card-content-left {
            text-align: left;
        }

        & .card-content-right {
            text-align: right;
        }
    }
}
