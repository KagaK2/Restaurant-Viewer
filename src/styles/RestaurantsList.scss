.restaurants-list{
    // This is the minimum width of a card.
    --grid-min-size: 25rem;

    display: grid;
    // The size of the cards will be flexible but bigger than or equal to 25rem.
    // Also making the cards-displaying grid more responsive.
    grid-template-columns: repeat(auto-fill, minmax(var(--grid-min-size), 1fr)); 
    gap: 1rem;
    margin: 0 1rem;

    // The size of the cards will be as wide as the content, but smaller than 25rem
    // For screen with width less than 420px so that the margins are displayed correctly.
    @media screen and (max-width: 420px){
        grid-template-columns: repeat(auto-fill, minmax(min-content, 25rem));
    }
}