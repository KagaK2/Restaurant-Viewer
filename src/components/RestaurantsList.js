import React from 'react';
import Card from './Card';
import '../styles/RestaurantsList.scss';

//This component takes an array of restaurants and returns a list of cards displaying the restaurants.
const RestaurantsList = ({restaurants}) => {
    return (
        //i is chosen since there is no unique restaurant id, name can be an alternative.
        <div className="restaurants-list">
            {restaurants.map((restaurant, i) => {
               return( 
                    <Card key={i} restaurant={restaurant} />
               );
            })}
        </div>
    )
}

export default RestaurantsList;