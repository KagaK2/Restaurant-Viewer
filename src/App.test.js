import Restaurant from './data/restaurants';
import {shallow} from 'enzyme';
import React from 'react';
import {sortRestaurantsAsc, sortRestaurantsDes} from './helpers';
import Card from './components/Card';
import RestaurantsList from './components/RestaurantsList';


describe('restaurant data', ()=>{
  test('is defined', () => {
    expect(Restaurant).toBeDefined();
  });
  
  test('has restaurants array in it', () => {
    expect(Restaurant.restaurants.length).toBeGreaterThanOrEqual(1);
  }); 


});

describe('restaurants got sorted alphabetically - Ascending', ()=> {
  const sortedArray = sortRestaurantsAsc(Restaurant.restaurants);

  test('the sorted array has the same number of items as the original array', ()=>{
    expect(sortedArray.length).toEqual(Restaurant.restaurants.length);
  });

  test('the sorted array has the right sorting order', () => {
    expect(sortedArray[0].name < sortedArray[1].name).toBeTruthy();
  })
  
  
})

describe('restaurants got sorted alphabetically - Descending', () => {
  const sortedArray = sortRestaurantsDes(Restaurant.restaurants);
  
  test('the sorted array has the same number of items as the original array', ()=>{
    expect(sortedArray.length).toEqual(Restaurant.restaurants.length);
  });

  test('the sorted array has the right sorting order', () => {
    expect(sortedArray[0].name > sortedArray[1].name).toBeTruthy();
  })
  
})

describe('Card component tests', ()=>{
  test('Card component with the first restaurant data got rendered correctly', () => {
    expect(shallow(<Card restaurant={Restaurant.restaurants[0]}/>)).toMatchSnapshot();
  })
})

describe('RestaurantsList tests', ()=>{
  const ascSortedArray = sortRestaurantsAsc(Restaurant.restaurants);
  const desSortedArray = sortRestaurantsDes(Restaurant.restaurants);
  test('Unsorted array of restaurants got rendered correctly', () => {
    expect(shallow(<RestaurantsList restaurants={Restaurant.restaurants} />)).toMatchSnapshot();
  })

  test('Ascending sorted array of restaurants got rendered correctly', () => {
    expect(shallow(<RestaurantsList restaurants={ascSortedArray} />)).toMatchSnapshot();
  })

  test('Descending sorted array of restaurants got rendered correctly', () => {
    expect(shallow(<RestaurantsList restaurants={desSortedArray} />)).toMatchSnapshot();
  })
  
})