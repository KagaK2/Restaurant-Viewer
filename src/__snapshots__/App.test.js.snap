// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card component tests Card component with the first restaurant data got rendered correctly 1`] = `ShallowWrapper {}`;

exports[`RestaurantsList tests Ascending sorted array of restaurants got rendered correctly 1`] = `ShallowWrapper {}`;

exports[`RestaurantsList tests Descending sorted array of restaurants got rendered correctly 1`] = `ShallowWrapper {}`;

exports[`RestaurantsList tests Unsorted array of restaurants got rendered correctly 1`] = `ShallowWrapper {}`;
